### PR TITLE
Make IOManager own resource context

### PR DESCRIPTION
**Pull request by dnfield**
Fixes flutter/flutter#24798

Successor to https://github.com/flutter/engine/pull/7243 - that approach seemed like it would ultimately fail, even though it has some similarities to this one.

This exposes an IOManager interface in `lib/ui` so that `runtime` and `lib/ui` can access the resource GrContext and SkiaUnrefQueue via the IOManager.  Ensures that the resource GrContext is created on the IO task runner.

This does not require changes to the Platform View implementation, so shouldn't affect Fuchsia the way the other one would have.

### CODE REVIEW

File: lib/ui/io_manager.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+#ifndef FLUTTER_LIB_UI_IO_MANAGER_H_
+#define FLUTTER_LIB_UI_IO_MANAGER_H_
+
+#include "flutter/flow/skia_gpu_object.h"
+#include "flutter/fml/memory/weak_ptr.h"
+#include "third_party/skia/include/gpu/GrContext.h"
+
+namespace blink {
+// Interface for methods that manage access to the resource GrConext and Skia
```
[the remark this answers is not in the document]
Comment (dnfield) [reply lib/ui/io_manager.h]: Thanks - fixed

File: lib/ui/ui_dart_state.cc [hunk 2]
```diff
@@ -114,7 +115,10 @@ fml::WeakPtr<SnapshotDelegate> UIDartState::GetSnapshotDelegate() const {
 }
 
 fml::WeakPtr<GrContext> UIDartState::GetResourceContext() const {
-  return resource_context_;
+  if (!io_manager_) {
+    return fml::WeakPtr<GrContext>();
```
Comment (chinmaygarde) [inline lib/ui/ui_dart_state.cc]: Nit: Since the type is already known, you can replace this line with just `return {};`

Comment (dnfield) [reply lib/ui/ui_dart_state.cc]: Done

**Comment by dnfield** [comment]
Note - updated some tests here for async image and updated the golden (looks like improvements in Skia's encoding caused this).

**Pull request merged**